### PR TITLE
DP-5: document tampering detection feature

### PR DESCRIPTION
**Pull request by barisere**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -648,16 +648,16 @@ The verification information consists of the following information:
    </td>
   </tr>
   <tr valign="top">
-   <td>Dispensations
+   <td>Dispenses
```
Comment (stephalba) [inline README.md:651]: Shouldn't this be "Events" (when looking at the example response?

File: README.md [hunk 2]
```diff
@@ -674,12 +674,48 @@ The verification information consists of the following information:
...
+         "hash": "5514842c6c8a86f88fd5f78037c77603ad67532bdd92334112e0ecf89f7742b3",
+         "parent_hash": "c2c9350309b99f9867f4c929ac22725cfe8f08719725ff6005d30aa085f4f4f6"
       }
    ]
 }
 ```
 
+##### 4.2.2.1. Tampering detection
+
+Verification will fail if the system detects that the prescription has been tampered. Here is how we detect tampering, briefly.
```
Comment (stephalba) [inline README.md:688]:
```suggestion
Verification will fail if the system detects that the prescription has been tampered with.
```

File: README.md [hunk 3]
```diff
@@ -674,12 +674,48 @@ The verification information consists of the following information:
...
+    "cancellable": false,
+    "hash": "aa9f0e8855bc25a5b45b5150e900cdde226ab0cccf31503d4a08783e7d46ec6b",
+    "signature": "<snip>",
+    "prescription_hash": "0x826ce404e00eddcaacb1019a4f759216cecca4ed9a6a16cc2d16e35d4a9e76bd"
+}
+```
+
+On verification, we check the following.
+
+1. The prescription's signature is valid for the prescription signing public key. This ensures that the QR code/URL data content (CHMED16A1 ePrescription data and signature) have not been modified.
```
Comment (stephalba) [inline README.md:710]:
```suggestion
1. The prescription's signature is validated against the public key. This ensures that the QR code/URL data content (CHMED16A1 E-Prescription data and signature) have not been modified.
```

File: README.md [hunk 4]
```diff
@@ -674,12 +674,48 @@ The verification information consists of the following information:
...
+2. The `create` event's `prescription_hash` field matches the hash of the CHMED16A1 ePrescription data.
```
Comment (stephalba) [inline README.md:711]:
```suggestion
2. The `create` event's `prescription_hash` field matches the hash of the CHMED16A1 E-Prescription data.
```

Comment (stephalba) [reply README.md:711]: Maybe also add why we do this?

File: README.md [hunk 5]
```diff
@@ -674,12 +674,48 @@ The verification information consists of the following information:
...
+3. The event chain has not been altered. We recalculate the events chain and compare it against both the last event and the audit log.
```
Comment (stephalba) [inline README.md:712]: Also add why we do this.

File: README.md [hunk 6]
```diff
@@ -674,12 +674,48 @@ The verification information consists of the following information:
...
+4. Each event signature is authentic. We check that each event's signing certificate links to a known root certificate authority that we control.
```
Comment (stephalba) [inline README.md:713]:
```suggestion
4. Each event signature is authentic. We check that each event's signing certificate links to a known root certificate.
```

Comment (stephalba) [reply README.md:713]: Also add why we do this.

File: README.md [hunk 7]
```diff
@@ -674,12 +674,48 @@ The verification information consists of the following information:
...
+
+If any of these checks fail, the verification fails.
+
+NOTE: the tampering checks are currently opt-in. To enable them, set the `ENABLE_EPRESCRIPTION_TAMPERING_CHECKS` environment variable to `true` when running the program.
```
Comment (stephalba) [inline README.md:717]:
```suggestion
```

Comment (stephalba) [reply README.md:717]: When we enable it, it should not be opt-in.

File: README.md (hunk 5 above)
Comment (barisere) [reply README.md:712]: That is explained in the first sentence. We are checking that the event log has not been altered by recalculating its final hash and comparing it to the audit log's.

File: README.md (hunk 1 above)
Comment (barisere) [reply README.md]: Changed

File: README.md (hunk 4 above)
Comment (barisere) [reply README.md]: Added in https://github.com/certifaction/hinsign-cli/pull/14/commits/31e5dacf3598b2218c235f850de02a8cd229e196

File: README.md (hunk 6 above)
Comment (barisere) [reply README.md]: Added in https://github.com/certifaction/hinsign-cli/pull/14/commits/31e5dacf3598b2218c235f850de02a8cd229e196

**Pull request merged**